src/context.zig

```diff
@@ -0,0 +1,457 @@
+        }
+        return init(allocator, deadline_ns);
+    }
+
+    /// Create a child context from parent with optional new deadline
+    pub fn withParent(parent: *const Context, deadline_ns: ?i128) !*Context {
+        const child = try parent.allocator.create(Context);
+        child.* = Context{
+            .allocator = parent.allocator,
+            .deadline_ns = if (deadline_ns) |d| blk: {
+                if (parent.deadline_ns) |parent_deadline| {
+                    break :blk @min(d, parent_deadline);
+                } else {
+                    break :blk d;
+                }
+            } else parent.deadline_ns,
+            .cancelled = false,
+            .cancel_reason = null,
+            .parent = parent,
+            .children = std.ArrayList(*Context).init(parent.allocator),
+            .mutex = std.Thread.Mutex{},
+        };
+
+        // Add to parent's children list (this would need proper synchronization in real use)
+        return child;
+    }
+
+    /// Clean up context resources
+    pub fn deinit(self: *Context) void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+
+        // Clean up children
+        for (self.children.items) |child| {
+            child.deinit();
+            self.allocator.destroy(child);
+        }
+        self.children.deinit();
+
+        if (self.cancel_reason) |reason| {
+            self.allocator.free(reason);
+        }
+    }
+
+    /// Internal const check for cancellation (no mutex, for parent checking)
+    fn isCancelledInternal(self: *const Context) bool {
+        // Check if explicitly cancelled
+        if (self.cancelled) return true;
+
+        // Check if parent is cancelled
+        if (self.parent) |parent| {
+            if (parent.isCancelledInternal()) return true;
+        }
+
+        return false;
+    }
+
+    /// Check if context is cancelled
+    pub fn isCancelled(self: *Context) bool {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+
+        return self.isCancelledInternal();
+    }
+
+    /// Check if deadline has been exceeded
+    pub fn isExpired(self: *Context) bool {
+        if (self.deadline_ns) |deadline| {
+            return time.nanoTimestamp() >= deadline;
+        }
+        return false;
+    }
+
+    /// Check if context is done (cancelled or expired)
+    pub fn isDone(self: *Context) bool {
+        return self.isCancelled() or self.isExpired();
+    }
+
+    /// Get the error reason for why context is done
+    pub fn err(self: *Context) ?ContextError {
+        if (self.isCancelled()) return ContextError.Cancelled;
+        if (self.isExpired()) return ContextError.DeadlineExceeded;
+        return null;
+    }
+
+    /// Cancel the context with optional reason
+    pub fn cancel(self: *Context, reason: ?[]const u8) !void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+
+        if (self.cancelled) return; // Already cancelled
+
+        self.cancelled = true;
```

Comment:
Again, maybe the atomic can be removed and an atomic operation can be use for `cancelled`, but if not, I don't think the mutex has to be locked beyond this point. Once `cancelled` is set to true, another thread cannot get passed the point(because of the `if (cancelled) return` check). I just don't like holding locks over allocations when avoidable.